session/pingpong/invoice_tracker.go

```diff
@@ -154,57 +137,153 @@ type InvoiceTrackerDeps struct {
+	it.invoiceLock.Lock()
+	defer it.invoiceLock.Unlock()
+
+	delete(it.invoicesSent, hex.EncodeToString(hashlock))
+}
+
+func (it *InvoiceTracker) getMarkedInvoice(hashlock []byte) (invoice sentInvoice, ok bool) {
+	it.invoiceLock.Lock()
+	defer it.invoiceLock.Unlock()
+	in, ok := it.invoicesSent[hex.EncodeToString(hashlock)]
+	return in, ok
+}
+
+func (it *InvoiceTracker) listenForExchangeMessages() error {
+	for {
+		select {
+		case pm := <-it.deps.ExchangeMessageChan:
+			err := it.handleExchangeMessage(pm)
+			if err != nil && err != ErrInvoiceExpired {
+				return err
+			}
+		case <-it.stop:
+			return nil
+		}
+	}
+}
+
+func (it *InvoiceTracker) handleExchangeMessage(pm crypto.ExchangeMessage) error {
+	invoice, ok := it.getMarkedInvoice(pm.Promise.Hashlock)
+	if !ok {
+		log.Debug().Msgf("consumer sent exchange message with missing expired hashlock %v, skipping", invoice.invoice.Hashlock)
+		return ErrInvoiceExpired
+	}
+
+	err := it.validateExchangeMessage(pm)
 	if err != nil {
-		return errors.Wrap(err, "could not get new agreement id")
+		return err
 	}
 
-	r := it.generateR()
-	invoice := crypto.CreateInvoice(agreementID, it.paymentInfo.GetPrice().Amount, 0, r)
-	invoice.Provider = it.providerID.Address
-	it.lastInvoice = lastInvoice{
-		invoice: invoice,
-		r:       r,
+	it.lastExchangeMessage = pm
+	it.markInvoicePaid(pm.Promise.Hashlock)
+	it.resetNotReceivedExchangeMessageCount()
+
+	needsRevealing := false
+	accountantPromise, err := it.deps.AccountantPromiseStorage.Get(it.deps.ProviderID, it.deps.AccountantID)
+	switch err {
+	case nil:
+		needsRevealing = !accountantPromise.Revealed
+	case ErrNotFound:
+		needsRevealing = false
+	default:
+		return errors.Wrap(err, "could not get accountant promise")
+	}
+
+	if needsRevealing {
+		err = it.deps.AccountantCaller.RevealR(accountantPromise.R, it.deps.ProviderID.Address, accountantPromise.AgreementID)
+		if err != nil {
+			log.Error().Err(err).Msg("Could not reveal R")
+			it.incrementAccountantFailureCount()
+			if it.getAccountantFailureCount() > it.deps.MaxAccountantFailureCount {
+				return errors.Wrap(err, "could not call accountant")
+			}
+			log.Warn().Msg("Ignoring accountant error, we haven't reached the error threshold yet")
+			return nil
+		}
+		it.resetAccountantFailureCount()
+		accountantPromise.Revealed = true
+		err = it.deps.AccountantPromiseStorage.Store(it.deps.ProviderID, it.deps.AccountantID, accountantPromise)
+		if err != nil {
+			return errors.Wrap(err, "could not store accountant promise")
+		}
+		log.Debug().Msg("Accountant promise stored")
+	}
+
+	err = it.deps.InvoiceStorage.StoreR(it.deps.ProviderID, it.agreementID, hex.EncodeToString(invoice.r))
+	if err != nil {
+		return errors.Wrap(err, "could not store r")
+	}
+
+	promise, err := it.deps.AccountantCaller.RequestPromise(pm)
+	if err != nil {
+		log.Warn().Err(err).Msg("Could not call accountant")
+
+		// TODO: handle this better
+		if strings.Contains(err.Error(), "400 Bad Request") {
```

Comment:
RequestPromise could return err which could be checked like:

```go
if err == accountant_caller_pkg.ErrBadRequest {
}
```

Reply:
this should in general unified with accountant via error codes, because 400 is not descriptive enough.

https://github.com/mysteriumnetwork/accountant/issues/50